### ISSUE TITLE
Segfault at sqsh__data_inode_modified_time()

### ISSUE THREAD

**Issue by brainstorm** [Gottox/sqsh-tools#298]
**Describe the bug**

I'm trying to unpack a "problematic" SquashFS image, extracted [from here](https://mega.nz/file/65xDmYZC#wQHG_cidg3rXzM6qvC71Dsa-ksIU7YOtplsmkXRNuRc) ... other tools seem to have issues with it too, see [this thread](https://x.com/braincode/status/1862979540420866472) for more context, but essentially (according to @devttys0):

> (...) Could be an issue with OOB/spare data in the raw flash dump.

> Interesting, carving out the SquashFS images and running unsquashfs manually provides a partial extraction, but encounters a decompression failure part-way through.

**To Reproduce**

Download the file above and extract the relevant SquashFS offset and size with r2:

```
[0x00000000]> s 0x2220000
[0x02220000]> wtf squashfs_part1.sqsh 22686932
INFO: Dumped 22686932 bytes from 0x02220000 into squashfs_part1.sqsh
```

As discussed with @trufae, there's issues in the r2 support as well (not sure I'm using it the right way though?):

```
% r2pm -s squash
sqsh-tools          SquashFS utilities
fs-squash           [r2-fs] read-only SquashFS read support

(... fs-squash does not compile ...)

% r2pm -ci sqsh-tools
% mkdir mnt && r2 squashfs_part1.sqsh
[0x00000000]> m ./mnt squashfs 0
ERROR: Invalid mountpoint
```

Then, running it outside of r2 via `sqsh-unpack` cmdline tool and against the sliced section it segfaults:

```
 % lldb sqsh-unpack squashfs_part1.sqsh
(lldb) target create "sqsh-unpack"
Current executable set to '/opt/homebrew/bin/sqsh-unpack' (arm64).
(lldb) settings set -- target.run-args  "squashfs_part1.sqsh"
(lldb) run
Process 12745 launched: '/opt/homebrew/bin/sqsh-unpack' (arm64)
Process 12745 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
    frame #0: 0x0000000100061fa4 libsqsh.1.dylib`sqsh__data_inode_modified_time(inode=0x0000000000000000) at inode_data.c:55:9
   52  	}
   53  	uint32_t
   54  	sqsh__data_inode_modified_time(const struct SqshDataInode *inode) {
-> 55  		return CX_LE_2_CPU32(inode->header.modified_time);
   56  	}
   57  	uint32_t
   58  	sqsh__data_inode_number(const struct SqshDataInode *inode) {
Target 0: (sqsh-unpack) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
  * frame #0: 0x0000000100061fa4 libsqsh.1.dylib`sqsh__data_inode_modified_time(inode=0x0000000000000000) at inode_data.c:55:9
    frame #1: 0x0000000100055720 libsqsh.1.dylib`sqsh_file_modified_time(inode=0x000000014bf04e90) at file.c:277:9
    frame #2: 0x0000000100002fcc sqsh-unpack`update_metadata(path=".init_enable_core", file=0x000000014bf04e90) at unpack.c:86:38
    frame #3: 0x0000000100002ec4 sqsh-unpack`extract_file_after(file=0x000000014bf04e90, stream=0x00000001fda2c0d0, d=0x0000600001ab0000, err=0) at unpack.c:190:7
    frame #4: 0x000000010005ef14 libsqsh.1.dylib`stream_worker(file=0x000000014bf04e90, iterator=0x0000000000000000, offset=0, data=0x000000014bf04e90, err=0) at file_ext.c:225:3
    frame #5: 0x000000010005f284 libsqsh.1.dylib`file_iterator_mt_cleanup(mt=0x000000014bf04e90, rv=0) at file_ext.c:100:2
    frame #6: 0x000000010005ede4 libsqsh.1.dylib`file_iterator_mt(mt=0x000000014bf04e90, file=0x000000014bf04c10, threadpool=0x000000014bf04600, cb=(libsqsh.1.dylib`stream_worker at file_ext.c:221), data=0x000000014bf04e90, rv=0) at file_ext.c:200:3
    frame #7: 0x000000010005ee9c libsqsh.1.dylib`sqsh_file_to_stream_mt(file=0x000000014bf04c10, threadpool=0x000000014bf04600, stream=0x00000001fda2c0d0, cb=(sqsh-unpack`extract_file_after at unpack.c:175), data=0x0000600001ab0000) at file_ext.c:265:2
    frame #8: 0x0000000100002be4 sqsh-unpack`extract_file(path=".init_enable_core", file=0x000000014bf04c10) at unpack.c:246:7
    frame #9: 0x0000000100002778 sqsh-unpack`extract_first_pass(path=".init_enable_core", type=SQSH_FILE_TYPE_FILE, file=0x000000014bf04c10) at unpack.c:320:10
    frame #10: 0x0000000100002714 sqsh-unpack`extract(path=".init_enable_core", file=0x000000014bf04c10, func=(sqsh-unpack`extract_first_pass at unpack.c:315)) at unpack.c:350:9
    frame #11: 0x000000010000326c sqsh-unpack`extract_from_traversal(target_path=".", iter=0x000000014bf04a50, func=(sqsh-unpack`extract_first_pass at unpack.c:315)) at unpack.c:385:8
    frame #12: 0x0000000100002914 sqsh-unpack`extract_all(target_path=".", base=0x000000014bf04710, func=(sqsh-unpack`extract_first_pass at unpack.c:315)) at unpack.c:427:8
    frame #13: 0x00000001000024bc sqsh-unpack`main(argc=2, argv=0x000000016fdfeb98) at unpack.c:535:8
    frame #14: 0x000000019858c274 dyld`start + 2840
(lldb)
```

**Expected behavior**

I expect all files to be extracted in the current directory (preserving original fs hierarchy, of course).

**Platform (please complete the following information):**
 - Operating System: OSX Sequoia
 - Version: 15.1.1

**Additional context**

https://github.com/hack-technicolor/hack-technicolor/issues/144#issuecomment-2401070332

**Comment by brainstorm** [comment]
Naïvely and quickly patching the segfaulting spots:

```diff
diff --git a/common/src/data/inode_data.c b/common/src/data/inode_data.c
index f7439d91..dae2fa29 100644
--- a/common/src/data/inode_data.c
+++ b/common/src/data/inode_data.c
@@ -40,6 +40,7 @@ sqsh__data_inode_type(const struct SqshDataInode *inode) {
 }
 uint16_t
 sqsh__data_inode_permissions(const struct SqshDataInode *inode) {
+       if (inode == NULL) { return 0; }
        return CX_LE_2_CPU16(inode->header.permissions);
 }
 uint16_t
@@ -52,6 +53,7 @@ sqsh__data_inode_gid_idx(const struct SqshDataInode *inode) {
 }
 uint32_t
 sqsh__data_inode_modified_time(const struct SqshDataInode *inode) {
+       if (inode == NULL) { return 0; }
        return CX_LE_2_CPU32(inode->header.modified_time);
 }
 uint32_t
```

Yields a great amount of correct files but others have all 0's on them (with seemingly correct filesizes)... also I just noticed that subsequent runs with the patch above are yielding different amounts of data instead of deterministically outputting the same bytes... kind of expected w/ stray null derefs, I guess.

**Comment by Gottox** [comment]
Uh, a malicious file! Apparently something slipped through. Awesome, I'll take a look later today! Thanks for providing the input!

**Comment by brainstorm** [comment]
Thanks to @wcampbell0x2a and @devttyS0 for getting me into reading about how NAND memories really work and their world of gotchas for extracting filesystems with "regular tools" such as binwalk et al:

https://blog.fraktal.fi/nand-on-my-watch-f307ac673d22
https://blog.fraktal.fi/ubi-or-not-ubi-insights-from-examining-50-nand-flashes-6e975ee273e5

I have the required parameters from [the NAND datasheet (page 39 mostly)](https://www.mouser.com/datasheet/2/671/micron_technology_mict_s_a0000866752_1-1759126.pdf), but I find it hard to find a "trusted" tool to reliably apply/fix those pesky ECC and block (re-)arrangements. 

I've tried a couple of them without much success so far, i.e, [FwHandler](https://github.com/LovenSar/FwHandler) does not seem to have good error control (no pun intended):

```
% python3 Fwhandler.py -f ../backhand/squashfs_part1.sqsh -e 4 -s 16384 -p 67108864 -u 1024 --tmp
(...)
Exist bad block.
Ecclen = 0, Spare = 4, Unit = 32, Page = 17408, Block = 0, Plane = 0
binFile = ../backhand/squashfs_part1.sqsh.fwtmp
Get File Size = 1377024 bytes
OutputFile: ../backhand/squashfs_part1.sqsh.fwhd.bin
Traceback (most recent call last):
  File "/Users/rvalls/dev/personal/fwhandler/Fwhandler.py", line 230, in <module>
    fileHandle(binFile, outputFile=binFile+".fwhd.bin", ecclen=ecclen, spare=spare, unit=unit,
  File "/Users/rvalls/dev/personal/fwhandler/Fwhandler.py", line 155, in fileHandle
    fileHandle(binFile=tmpFile, outputFile=outputFile, ecclen=0, spare=ecclen, unit=unit,
  File "/Users/rvalls/dev/personal/fwhandler/Fwhandler.py", line 110, in fileHandle
    sh0, sh1, sh2 = sb[0], sb[1], sb[2]
                    ~~^^^
IndexError: index out of range
```

Or probably there's indeed a few badblocks in the dump... hard to tell and trust without staring at the bytes and datasheet and keep trying via informed guesses.

So TL;DR, while your software segfaults, the rabbit hole runs deeper to properly decode this NAND dump.

**Comment by Gottox** [comment]
Quicker way to reproduce without radare:

1. Download file
2. `build/tools/sqsh-unpack ~/Downloads/Vodafone-WifiHub-Micron_MT29F1G08ABAEA_00-07FFFFFF.bin -o 0x2220000`

It doesn't segfault with `sqsh-ls -rl`, which is a bit strange.

**Comment by brainstorm** [comment]
> It doesn't segfault with sqsh-ls -rl, which is a bit strange.

I noticed non-deterministic behavior during my tests w/ `sqsh-unpack` (perhaps UB?), sometimes it spew "Bus Error" and didn't output as many files as with the immediately previous run, different contents in some files, etc... Did you run `valgrind` to scan for memleaks and other issues?

**Comment by Gottox** [comment]
@brainstorm That's actually possible, as sqsh-unpack runs multithreaded. This happens with the segfault or in general?

**Comment by Gottox** [comment]
The issue is fixed with [v1.5.2](https://github.com/Gottox/sqsh-tools/releases/tag/v1.5.2).

**Comment by brainstorm** [comment]
Is it really fixed for good? I just git pulled, `meson setup wipe`, `meson compile`, `meson install` and I'm getting the "bus error" (at the end)... also the "Compression decompress" is meant to be an error or nomal message meaning "Successfully decompressed"?

```
$ ../sqsh-tools/build/tools/sqsh-unpack squashfs_part1.sqsh

(...)
usr/lib/lua/transformer/shared/VendorConfig.lua: Compression decompress
usr/lib/libwlctl.so: Compression decompress
usr/lib/lua/transformer/shared/VendorLog.lua: Compression decompress
usr/lib/libvlanctl.so: Compression decompress
usr/lib/lua/tch/inet.lua: Compression decompress
usr/lib/lua/transformer/shared/WLANConfigurationCommon.lua: Compression decompress
usr/lib/lua/transformer/eventsource/ubus.lua: Compression decompress
usr/lib/lua/transformer/shared/atmpinghelper.lua: Compression decompress
usr/lib/lua/todfw/init.lua: Compression decompress
usr/lib/lua/transformer/shared/banktable.lua: Compression decompress
usr/lib/lua/transformer/mapper/ucihelper.lua: Compression decompress
usr/lib/lua/transformer/logger.lua: Compression decompress
usr/lib/lua/transformer/shared/cmdhelper.lua: Compression decompress
usr/lib/lua/transformer/api.lua: Compression decompress
usr/lib/lua/tch/simplecrypto.so: Compression decompress
usr/lib/lua/transformer/shared/custo/wifisecurity.lua: Compression decompress
usr/lib/lua/socket/headers.lua: Compression decompress
usr/lib/lua/transformer/msg.lua: Compression decompress
usr/lib/lua/transformer/mapper/ubus.lua: Compression decompress
usr/lib/lua/transformer/shared/ddns.lua: Compression decompress
usr/lib/lua/socket/http.lua: Compression decompress
usr/lib/lua/tch/tableprint.lua: Compression decompress
usr/lib/lua/transformer/eventsource/uci.lua: Compression decompress
usr/lib/lua/transformer/mapper/nwcommon.lua: Compression decompress
usr/lib/lua/transformer/shared/devicehelper.lua: Compression decompress
usr/lib/lua/transformer/mapper/uci_1to1.lua: Compression decompress
usr/lib/lua/tch.so: Compression decompress
usr/lib/lua/transformer/shared/dhcp.lua: Compression decompress
usr/lib/lua/socket/smtp.lua: Compression decompress
usr/lib/lua/transformer/mapper/ucilist.lua: Compression decompress
usr/lib/lua/transformer/navigation.lua: Compression decompress
usr/lib/lua/transformer/shared/firewall_helper.lua: Compression decompress
usr/lib/lua/transformer/msg_encode.lua: Compression decompress
usr/lib/lua/socket.lua: Compression decompress
usr/lib/lua/transformer/shared/bandsteerhelper.lua: Compression decompress
usr/lib/lua/transformer/msg_decode.lua: Compression decompress
usr/lib/lua/transformer/shared/fon_helper.lua: Compression decompress
usr/lib/lua/transformer/shared/ipdiagnosticshelper.lua: Compression decompress
usr/lib/lua/transformer/shared/ippinghelper.lua: Compression decompress
usr/lib/lua/transformer/shared/mobileconnection.lua: Compression decompress
usr/lib/lua/transformer/lock.lua: Compression decompress
usr/lib/lua/transformer/commitapply.lua: Compression decompress
usr/lib/lua/transformer/mapper/command.lua: Compression decompress
usr/lib/lua/transformer/shared/intfdiaghelper.lua: Compression decompress
usr/lib/mmpbx/libmmpbxrvsipnet.so: Compression decompress
usr/sbin/ebtables: Compression decompress
usr/sbin/fileprofiler: Compression decompress
zsh: bus error  ../sqsh-tools/build/tools/sqsh-unpack squashfs_part1.sqsh
```

**Comment by Gottox** [comment]
Could you provide more system information here? I cannot reproduce the bus error unfortunately.

**Comment by brainstorm** [comment]
> Could you provide more system information here? I cannot reproduce the bus error unfortunately.

```
Darwin MacBookPro.internal 24.1.0 Darwin Kernel Version 24.1.0: Thu Oct 10 21:05:14 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T8103 arm64 arm Darwin
```

My apologies, I just realised that `v1.5.2` hasn't been merged to `master` (wrong assumption).

Also, while I'm here, the optional flags confuse me a bit: what's the difference of `[PATH]`
 and `[TARGET DIR]`?:

```
rvalls@MacBookPro vodafone_router % sqsh-unpack -h
usage: sqsh-unpack [-o OFFSET] [-cVRe] FILESYSTEM [PATH] [TARGET DIR]
       sqsh-unpack -v
rvalls@MacBookPro vodafone_router % sqsh-unpack squashfs_part1.sqsh extracted
extracted: No such file or directory
rvalls@MacBookPro vodafone_router % sqsh-unpack squashfs_part1.sqsh ./extracted
./extracted: No such file or directory
rvalls@MacBookPro vodafone_router % ls -alh extracted
total 0
drwxr-xr-x   2 rvalls staff   64 Dec 10 14:45 .
drwxr-xr-x 231 rvalls staff 7.3K Dec 10 14:45 ..
```

**Comment by Gottox** [comment]
Okay, about the bus error: I now have access to a MacOS system to debug this issue. It would be awesome if you could open a new bug report for this error.

**Comment by brainstorm** [comment]
> Okay, about the bus error: I now have access to a MacOS system to debug this issue. It would be awesome if you could open a new bug report for this error.

No need, the bug is fixed as I mentioned in `no bus error, still generates 0-byte-filled files though.`, thank you!
